cranelift/filetests/src/function_runner.rs

```diff
@@ -1,96 +1,389 @@
+        Self {
+            page,
+            signature,
+            trampoline,
+        }
+    }
+
+    /// Return a pointer to the compiled code.
+    pub fn as_ptr(&self) -> *const u8 {
+        self.page.as_ptr()
+    }
+
+    /// Call the [CompiledFunction], passing in [DataValue]s using a compiled [Trampoline].
+    pub fn call(&self, arguments: &[DataValue]) -> Vec<DataValue> {
+        let mut values = UnboxedValues::make_arguments(arguments, &self.signature);
+        let arguments_address = values.as_mut_ptr();
+        let function_address = self.as_ptr();
+
+        let callable_trampoline: fn(*const u8, *mut u128) -> () =
+            unsafe { mem::transmute(self.trampoline.as_ptr()) };
+        callable_trampoline(function_address, arguments_address);
+
+        values.collect_returns(&self.signature)
+    }
+}
+
+/// A container for laying out the [ValueData]s in memory in a way that the [Trampoline] can
+/// understand.
+struct UnboxedValues(Vec<u128>);
+
+impl UnboxedValues {
+    /// The size in bytes of each slot location in the allocated [DataValue]s. Though [DataValue]s
+    /// could be smaller than 16 bytes (e.g. `I16`), this simplifies the creation of the [DataValue]
+    /// array and could be used to align the slots to the largest used [DataValue] (i.e. 128-bit
+    /// vectors).
+    const SLOT_SIZE: usize = 16;
+
+    /// Build the arguments vector for passing the [DataValue]s into the [Trampoline]. The size of
+    /// `u128` used here must match [Trampoline::SLOT_SIZE].
+    pub fn make_arguments(arguments: &[DataValue], signature: &ir::Signature) -> Self {
+        assert_eq!(arguments.len(), signature.params.len());
+        let mut values_vec = vec![0; max(signature.params.len(), signature.returns.len())];
+
+        // Store the argument values into `values_vec`.
+        for ((arg, slot), param) in arguments.iter().zip(&mut values_vec).zip(&signature.params) {
+            if arg.ty() != param.value_type {
+                panic!("argument type mismatch");
+            }
+            unsafe {
+                Self::write_value_to(arg, slot);
+            }
+        }
+
+        Self(values_vec)
+    }
+
+    /// Return a pointer to the underlying memory for passing to the trampoline.
+    pub fn as_mut_ptr(&mut self) -> *mut u128 {
+        self.0.as_mut_ptr()
+    }
+
+    /// Collect the returned [DataValue]s into a [Vec]. The size of `u128` used here must match
+    /// [Trampoline::SLOT_SIZE].
+    pub fn collect_returns(&self, signature: &ir::Signature) -> Vec<DataValue> {
+        assert!(self.0.len() >= signature.returns.len());
+        let mut returns = Vec::with_capacity(signature.returns.len());
+
+        // Extract the returned values from this vector.
+        for (slot, param) in self.0.iter().zip(&signature.returns) {
+            unsafe {
+                returns.push(Self::read_value_from(slot, param.value_type));
```

Comment:
nit: can we put the `unsafe` within the `push`, to make it clear that the push itself isn't an issue?